### PR TITLE
Generalize BSI model and add two variants

### PR DESCRIPTION
**Pull request by n01r**
1. The **_barrier suppression ionization_** model is hereby generalized to calculate threshold field strengths from given ionization energies

2. `BSI` has been renamed to `BSIHydrogenLike` to emphasize that the model has been derived from the hydrogen atom and was extrapolated to hydrogen-like and other atoms by using their ionization potentials.
It should be used as a first but crude model to see the effects of ionization in a given simulation setup.

3. Two other BSI variants have been added:
    - `BSIStarkShifted` was developed for hydrogen and hydrogen-like atoms and accounts for the Stark upshift of the ionization potential in a strong external electric field.    
    (2010, Bauer, Mulser: _High power laser-matter interaction_)
    - `BSIEffectiveZ` is my own attempt to generalize the BSI model to non-hydrogen-like atoms. Effective atomic numbers for each shell electron are used to account for an inner shell shielding effects. Thereby the derivation from the hydrogen problem might get a bit more justified.

***Side note:*** The Wiki has been updated with a [step-by-step example](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/Ionization-in-PIConGPU#howto-set-up-a-parameter-set-with-activated-ionization) on how to create a parameter set with ionization activated

### CODE REVIEW

**Comment by bussmann** [comment]
Hi, how about adding some validation to these models and also test cases (variation in atomic and laser parameters) that automatically test against known results and illustrate the difference in the models?

**Comment by n01r** [comment]
Absolutely! I just thought that I would add the models first so that it becomes transparent what is being tested. This was also still a remnant from my thesis which (I was certain before) had already found its way into `dev`. Since it didn't I opened it for reviewing now. 

One of these model variations is my own little idea, though. (At least I haven't seen it anywhere else before.) So I am interested in how well it fares in comparison to the others.

File: src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp [hunk 1]
```diff
@@ -64,13 +65,16 @@ namespace ionization
...
 
-            const float_X protonNumber = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
-            float_X chargeState = attribute::getChargeState(parentIon);
-
-            uint32_t cs = math::float2int_rd(chargeState);
+            const float_X protonNumber  = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            float_X chargeState         = attribute::getChargeState(parentIon);
+            uint32_t cs                 = math::float2int_rd(chargeState);
+            /* ionization potential in atomic units */
+            const float_X iEnergy       = GetIonizationEnergies<ParticleType>::type()[cs];
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: If the ion is fully ionized and `chargeState` is `protonNumber` this results in an invalid memory access. 
Ionization energies are only defined for the index `[0,protonNumber-1]`

Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: Argh! I should move the condition 
```C++
    if (chargeState < protonNumber)
```
before the definition of the ionization energy! Also the `GetIonizationEnergies` trait should receive a fallback for this!

Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: Yes

```C++
            const float_X protonNumber  = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
            float_X chargeState         = attribute::getChargeState(parentIon);
            uint32_t cs                 = math::float2int_rd(chargeState);
            /* ionization potential in atomic units */
           if(chargeState < protonNumber)
           {
              const float_X iEnergy       = GetIonizationEnergies<ParticleType>::type()[cs];
              /* critical field strength in atomic units */
              float_X critField           = math::pow(iEnergy,float_X(2.)) / protonNumber;

              /* ionization condition */
              if (math::abs(eField) / ATOMIC_UNIT_EFIELD >= critField)
              {
                  /* set new particle charge state */
                  parentIon[boundElectrons_] -= float_X(1.0);
              }
            }
```

File: src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp [hunk 2]
```diff
@@ -64,13 +65,16 @@ namespace ionization
...
+            /* critical field strength in atomic units */
+            float_X critField           = math::pow(iEnergy,float_X(2.)) / protonNumber;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: please use `math::pow(iEnergy,2)` or `iEnergy*iEnergy`.

The current used pow function with two floats is very slow.

File: src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp (hunk 1 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: Already fixed and pushed to the branch in all three versions of the BSI model. Thx @psychocoderHPC for spotting that!

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+            /* verify that ion is not completely ionized */
+            if (chargeState < protonNumber)
+            {
+                uint32_t cs                 = math::float2int_rd(chargeState);
+                /* ionization potential in atomic units */
+                const float_X iEnergy       = GetIonizationEnergies<ParticleType>::type()[cs];
+                const float_X ZEff          = GetEffectiveAtomicNumbers<ParticleType>::type()[cs];
+                /* critical field strength in atomic units */
+                float_X critField           = math::pow(iEnergy,float_X(2.)) / ZEff;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: please use the `math::pow(float,int)` or `value*value` else it is very slow

File: src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp [hunk 4]
```diff
@@ -64,16 +65,23 @@ namespace ionization
         operator()( const BType bField, const EType eField, ParticleType& parentIon )
         {
 
-            const float_X protonNumber = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
-            float_X chargeState = attribute::getChargeState(parentIon);
+            const float_X protonNumber  = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            float_X chargeState         = attribute::getChargeState(parentIon);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIHydrogenLike/AlgorithmBSIHydrogenLike.hpp]: please remove the spaces before `=`. We not align `=`.

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+         * \param eField electric field value at t=0
+         * \param parentIon particle instance to be ionized with position at t=0 and momentum at t=-1/2
+         */
+        template<typename EType, typename BType, typename ParticleType >
+        HDINLINE void
+        operator()( const BType bField, const EType eField, ParticleType& parentIon )
+        {
+
+            const float_X protonNumber  = GetAtomicNumbers<ParticleType>::type::numberOfProtons;
+            float_X chargeState         = attribute::getChargeState(parentIon);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: please remove spaces before `=`

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+            /* verify that ion is not completely ionized */
+            if (chargeState < protonNumber)
+            {
+                uint32_t cs                 = math::float2int_rd(chargeState);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: same here no `=` alignment pleace

File: src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    struct AlgorithmBSIStarkShifted
+    {
+
+        /** Functor implementation
+         *
+         * \tparam EType type of electric field
+         * \tparam BType type of magnetic field
+         * \tparam ParticleType type of particle to be ionized
+         *
+         * \param bField magnetic field value at t=0
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp]: Is `t=0` meaning the field at time step zero? Or the field strange at the particle position?

File: src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            float_X chargeState = attribute::getChargeState(parentIon);
+
+            /* verify that ion is not completely ionized */
+            if (chargeState < protonNumber)
+            {
+                uint32_t cs = math::float2int_rd(chargeState);
+                /* ionization potential in atomic units */
+                const float_X iEnergy = GetIonizationEnergies<ParticleType>::type()[cs];
+                /* critical field strength in atomic units */
+                float_X critField = (math::sqrt(float_X(2.))-float_X(1.)) * math::pow(iEnergy,float_X(3./2.));
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp]: This equation not use the number of protons `protonNumber` but our second implemented modules does it. Is this correct?

File: src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp (hunk 7 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp]: Ah, this is just a remnant from the pusher documentation. It means t=0 with respect to the current time step.
In the pushers you also sometimes get data from t=1/2 or t=-1/2.

File: src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp (hunk 8 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp]: This is also correct but this variant of the BSI model was only developed for Hydrogen or hydrogenlike ions (meaning if protonNumber = Z then a hydrogenlike ion has chargeState = Z-1). It is rather untested, yet, how good the model describes the ionization of other charge states.

File: src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp (hunk 7 above)
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSIStarkShifted/AlgorithmBSIStarkShifted.hpp]: Could you please add that it means the time relative to the timestep.

**Comment by ax3l** [comment]
@n01r can you please add the results of the three plane wave tests we taked about yesterday?
I still want to know the origin of the straight/staircase front in https://github.com/ComputationalRadiationPhysics/picongpu/issues/1469#issuecomment-221021911

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include "simulation_defines.hpp"
+#include "particles/traits/GetIonizationEnergies.hpp"
+#include "particles/traits/GetAtomicNumbers.hpp"
+#include "particles/traits/GetEffectiveAtomicNumbers.hpp"
+#include "traits/attribute/GetChargeState.hpp"
+#include "algorithms/math/floatMath/floatingPoint.tpp"
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: already included in `libPMacc/include/algorithms/math.hpp` which is included in PIConGPU's `simulation_types.hpp` which is already present in PIConGPU's `simulation_defines.hpp`

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: already included in PIConGPU's `simulation_defines.hpp`

**Comment by ax3l** [comment]
I don't like that much that all the `_Impl.hpp` are essentially identical and copy-pasted. Can you just create a common class with methods caching and preparation methods all field ionization methods need? Should be enough to only specialize the `operator()`.

**Comment by ax3l** [comment]
next time: please refactor existing code first (BSI Hydro) and then in a 2nd PR add new functionality (feature).

File: src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+         *
+         * \param bField magnetic field value at t=0
+         * \param eField electric field value at t=0
+         * \param parentIon particle instance to be ionized with position at t=0 and momentum at t=-1/2
+         *
+         * and "t" being with respect to the current time step (on step/half a step backward/-""-forward)
+         */
+        template<typename EType, typename BType, typename ParticleType >
+        HDINLINE void
+        operator()( const BType bField, const EType eField, ParticleType& parentIon )
```
Comment (ax3l) [inline src/picongpu/include/particles/ionization/byField/BSIEffectiveZ/AlgorithmBSIEffectiveZ.hpp]: unused var `bField`

**Comment by ax3l** [comment]
To make the new fields a bit less verbose: please just use the same method for field caching where applicable. E.g., write a `cacheEField` (BSI, BSI_EffZ, ADK) and `cacheEBField` (Shift) and `cacheDensTemp` (TF) ...

**Comment by ax3l** [comment]
the updates look good to me.
I co-assigned @psychocoderHPC in case he might see some further points, otherwise we can merge :)

**Comment by n01r** [comment]
I could maybe do a `git reset --soft` and reshape the commit structure before unless you deem that unnecessary.
I also started working on writing the `cacheEField`-like methods but than can be a different pull request.

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 12]
```diff
@@ -61,9 +65,48 @@ namespace ionization
...
+     * - accounts for stark upshift of ionization energy but only covers the
+     *   hydrogenlike ions originally
+     *
+     * \tparam T_DestSpecies electron species to be created
+     */
+    template<typename T_DestSpecies>
+    struct BSIStarkShifted
     {
-        typedef BSI_Impl<T_DestSpecies> type;
+        typedef particles::ionization::AlgorithmBSIStarkShifted T_IonizationAlgorithm;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: please remove the ` T_` in  `T_IonizationAlgorithm`

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 13]
```diff
@@ -61,9 +65,48 @@ namespace ionization
      * \see speciesDefinition.param
      */
     template<typename T_DestSpecies>
-    struct BSI
+    struct BSIHydrogenLike
+    {
+        typedef particles::ionization::AlgorithmBSIHydrogenLike T_IonizationAlgorithm;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: please remoe the `T_`

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 14]
```diff
@@ -61,9 +65,48 @@ namespace ionization
...
+     * - tries to account for electron shielding by issuing a lookup table of
+     *   effective atomic numbers for each electron shell @see ionizerConfig.param
+     * - unvalidated and still in development
+     *
+     * \tparam T_DestSpecies electron species to be created
+     */
+    template<typename T_DestSpecies>
+    struct BSIEffectiveZ
+    {
+        typedef particles::ionization::AlgorithmBSIEffectiveZ T_IonizationAlgorithm;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: remove `T_`

File: src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp [hunk 15]
```diff
@@ -20,15 +20,14 @@
...
 
-#include "pmacc_types.hpp"
 #include "simulation_defines.hpp"
+#include "particles/traits/GetIonizationEnergies.hpp"
 #include "particles/traits/GetAtomicNumbers.hpp"
 #include "traits/attribute/GetChargeState.hpp"
-#include "algorithms/math/floatMath/floatingPoint.tpp"
 
-/** \file AlgorithmBSI.hpp
+/** \file AlgorithmBSIStarkShifted.hpp
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp]: please remove this

File: src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp [hunk 16]
```diff
@@ -42,38 +41,44 @@ namespace particles
 namespace ionization
 {
 
-    /** \struct AlgorithmBSI
+    /** \struct AlgorithmBSIStarkShifted
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp]: please remove `\struct AlgorithmBSIStarkShifted`

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 17]
```diff
@@ -28,7 +28,7 @@ namespace particles
 {
 namespace ionization
 {
-    /** \struct BSI_Impl
+    /** \struct BSIHydrogenLike_Impl
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: please remove  `\struct BSIHydrogenLike_Impl`

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 18]
```diff
@@ -61,9 +65,48 @@ namespace ionization
...
      */
     template<typename T_DestSpecies>
-    struct BSI
+    struct BSIHydrogenLike
+    {
+        typedef particles::ionization::AlgorithmBSIHydrogenLike T_IonizationAlgorithm;
+        typedef BSI_Impl<T_IonizationAlgorithm, T_DestSpecies> type;
+    };
+
+        /** \struct BSIEffectiveZ
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: remove `\struct ...`

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 19]
```diff
@@ -61,9 +65,48 @@ namespace ionization
...
+     * \tparam T_DestSpecies electron species to be created
+     */
+    template<typename T_DestSpecies>
+    struct BSIEffectiveZ
+    {
+        typedef particles::ionization::AlgorithmBSIEffectiveZ T_IonizationAlgorithm;
+        typedef BSI_Impl<T_IonizationAlgorithm, T_DestSpecies> type;
+    };
+
+        /** \struct BSIStarkShifted
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: remove `\struct ...`

File: src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp [hunk 20]
```diff
@@ -42,38 +41,44 @@ namespace particles
...
          *
          * \tparam EType type of electric field
-         * \tparam BType type of magnetic field
          * \tparam ParticleType type of particle to be ionized
          *
-         * \param bField magnetic field value at t=0
          * \param eField electric field value at t=0
          * \param parentIon particle instance to be ionized with position at t=0 and momentum at t=-1/2
          */
         template<typename EType, typename BType, typename ParticleType >
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp]: `BType` is not needed anymore

File: src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp [hunk 21]
```diff
@@ -167,22 +155,18 @@ namespace ionization
...
                 const fieldSolver::numericalCellType::traits::FieldPosition<FieldE> fieldPosE;
                 ValueType_E eField = Field2ParticleInterpolation()
                     (cachedE.shift(localCell).toCursor(), pos, fieldPosE());
-                /*                     and B-field on the particle position */
-                const fieldSolver::numericalCellType::traits::FieldPosition<FieldB> fieldPosB;
-                ValueType_B bField = Field2ParticleInterpolation()
-                    (cachedB.shift(localCell).toCursor(), pos, fieldPosB());
 
                 /* define number of bound macro electrons before ionization */
                 float_X prevBoundElectrons = particle[boundElectrons_];
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp]: This observation  of the `ionizeAlgo` should be design in a clear interface. This avoid slow memory access `newMacroElectrons = prevBoundElectrons - particle[boundElectrons_];` to get the number of new electrons.

should be:
```C++
newMacroElectrons = ionizeAlgo( eField, particle);
```

File: src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp [hunk 22]
```diff
@@ -42,38 +41,44 @@ namespace particles
...
+                /* ionization potential in atomic units */
+                const float_X iEnergy = GetIonizationEnergies<ParticleType>::type()[cs];
+                /* critical field strength in atomic units */
+                float_X critField = (math::sqrt(float_X(2.))-float_X(1.)) * math::pow(iEnergy,float_X(3./2.));
+
+                /* ionization condition */
+                if (math::abs(eField) / ATOMIC_UNIT_EFIELD >= critField)
+                {
+                    /* set new particle charge state */
+                    parentIon[boundElectrons_] -= float_X(1.0);
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp]: offline discussed: This algorithm should only calculate how many electrons are created but should not manipulate the particle in the same moment.

Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp]: You also addressed the topic of #963 by expressing your discomfort about the `Algorithm[ADK|...].hpp` and `IonizationAlgorithm`. Feel free to add your suggestions to the issue. :+1:

File: src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp (hunk 21 above)
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp]: I will do a small change to that but refactoring the interface would also require me to change `ionization.hpp` and `WriteElectronIntoFrame` from `ionizationMethods.hpp`. That should be a separate PR.

Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI_Impl.hpp]: While I am at it I should make use of @Heikman's generalization (#1353) of my particle creation kernel, anyway, right?

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
+//#include "particles/ionization/byField/BSI/AlgorithmBSIHydrogenLike.hpp"
```
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: Remove unnecessary includes, you dunce! :stuck_out_tongue:

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+//#include "particles/ionization/byField/BSI/AlgorithmBSIEffectiveZ.hpp"
```
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: ... and here

File: src/picongpu/include/particles/ionization/byField/BSI/BSI.def [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+//#include "particles/ionization/byField/BSI/AlgorithmBSIStarkShifted.hpp"
```
Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/BSI/BSI.def]: ... and here!

**Comment by n01r** [comment]
As follow-ups we will have these two features:
- [topic-fieldCachingMethods](https://github.com/n01r/picongpu/tree/topic-fieldCachingMethods)   
 __*current progress*__: *E-Field caching method works, B-Field should be similar and both could maybe be unified to a single templated method*
- [topic-addIonModelADKForCircPol](https://github.com/n01r/picongpu/tree/topic-addIonModelADKForCircPol)   
*Adds a second ADK implementation for circularly polarized light. Both ADK formulas are from the same scientific source. (*Delone, Krainov: Tunneling and barrier-suppression ionization of atoms and ions in a laser radiation field., 1998*)*

**Comment by ax3l** [comment]
well done, thanks! :sparkles:

**Pull request merged**